### ISSUE TITLE
Build on FreeBSD 10.4

### ISSUE THREAD

**Issue by jjpprrrr** [lkl/linux#443]
After building LKL on FreeBSD 10.4 successfully, I would like to share some of my experiences.

1. The README still suggests using gcc49 which is a really old version. The latest gcc works just fine. I also installed some other GNU version of libs:

```sh
pkg install binutils gcc gnubc gmade gsed coreutils argp-standalone bison flex python
```

Note: binutils provides gnustat to replace stat, as in issue #368 .
2. in /scripts/link-vmlinux.sh, replace "stat" with "gnustat". I see that line 55-60 in /tools/lkl/Makefile is trying to address this issue, but it is not working for me.
3. in /scripts/headers_install.sh, replace "sed" with "gsed".
4. in /tools/lkl/include/lkl_host.h:25, comment out the definition of struct iovec because it is already defined somewhere. However, I did not see the same problem when building on Linux.
```C
 22 #ifdef LKL_HOST_CONFIG_POSIX
 23 #include <sys/uio.h>
 24 #else
 25 /*struct iovec {
 26         void *iov_base;
 27         size_t iov_len;
 28 };*/
 29 #endif
```

5. Sometimes compiling with "-fPIC" flag would help, though I did compile once without this flag.

I see that building hijack is disabled for BSD. Will there be a *BSD port of hijack in the future? Thanks!
```
ifneq ($(LKL_HOST_CONFIG_BSD),y)
libs-$(LKL_HOST_CONFIG_POSIX) += lib/hijack/liblkl-hijack
endif
```

**Comment by thehajime** [comment]
Thanks @jjpprrrr for the detailed report.
It would be nice if you could provide a patch (and pull request) for your improvement.

> After building LKL on FreeBSD 10.4 successfully, I would like to share some of my experiences.
> 
> The README still suggests using gcc49 which is a really old version. The latest gcc works just fine. I also installed some other GNU version of libs:
> pkg install binutils gcc gnubc gmade gsed coreutils argp-standalone bison flex python
> Note: binutils provides gnustat to replace stat, as in issue #368 .

indeed.

> 2. in /scripts/link-vmlinux.sh, replace "stat" with "gnustat". I see that line 55-60 in /tools/lkl/Makefile is trying to address this issue, but it is not working for me.

which version of lkl are you using ?

with the following commit on my FreeBSD11 (`lkldocker/circleci-freebsd11-x86_64:0.4` docker image), I don't have to edit link-vmlinux.sh (tools/lkl/Makefile correctly workaround this replacement).

```
commit 5e56fc93defeea368c48942fb589fcbe84af3b58 (HEAD -> master, origin/master, origin/HEAD)
Merge: ba78f51b8820 8b9f015c4261
Author: Hajime Tazaki <thehajime@gmail.com>
Date:   Wed Jun 13 07:43:19 2018 +0900

    Merge pull request #439 from libos-nuse/fix-qdisc-conf
    
    Fix qdisc configuration issues
```

> 3. in /scripts/headers_install.sh, replace "sed" with "gsed".

you can create a symlink of `gsed` as `sed` in tools/lkl/bin as gnustat does.

> 4. in /tools/lkl/include/lkl_host.h:25, comment out the definition of struct iovec because it is already defined somewhere. However, I did not see the same problem when building on Linux.
> 
>  22 #ifdef LKL_HOST_CONFIG_POSIX
>  23 #include <sys/uio.h>
>  24 #else
>  25 /*struct iovec {
>  26         void *iov_base;
>  27         size_t iov_len;
>  28 };*/
>  29 #endif

I don't have this issue (dup symbol of struct iovec) with the commit (above).

> Sometimes compiling with "-fPIC" flag would help, though I did compile once without this flag.
> I see that building hijack is disabled for BSD. Will there be a *BSD port of hijack in the future? Thanks!
> 
> ifneq ($(LKL_HOST_CONFIG_BSD),y)
> libs-$(LKL_HOST_CONFIG_POSIX) += lib/hijack/liblkl-hijack
> endif

I have an WIP branch on this support but it's not as simple as I expected..  I'm almost convinced to go with frankenlibc (in my other repository) for FreeBSD support.

https://github.com/libos-nuse/lkl-linux/commit/7c7f78b9ec4cc5ec4aa073ed134c0ce669d69a5a

**Comment by thehajime** [comment]
> I have an WIP branch on this support but it's not as simple as I expected.
> 
> For hijack, I wonder if the Linux Compatibility Emulation of FreeBSD would provide a temporary workaround. I was able to run the Linux binary of /tools/lkl/tests/boot on FreeBSD and passed all tests.

Linuxlator in FreeBSD does the opposite translation (Linux syscall to FreeBSD syscall) while this hijack library does the other way around.  So I thought it doesn't help to our case.

**Comment by thehajime** [comment]
Fixed by #444. Thank @jjpprrrr
